### PR TITLE
Add MSAA mode for Quest

### PR DESCRIPTION
**Pull request by BastiaanOlij**
Due to performance issues we want to do a single pass render on the Quest. The Quest has a wonderful feature where we can directly render into the textures supplied by the Oculus SDK and have MSAA applied as long as we set our frame buffer up correctly.

There was a previous hacked version that attempted to do the same but broke the Rift plugin and potentially also the OpenXR drivers.

This PR introduces two new MSAA modes specific to the Quest and only work with the GLES2 driver (the GLES3 drivers effects setup make this approach impossible as far as I can tell so they just behave as the normal MSAA 2x and 4x modes). 

When these modes are chosen Godot does not build its MSAA buffers instead setting up the FBO for the external texture correctly. 
If these new modes are used on anything but Quest hardware MSAA remains turned off.

As we're already rendering to the external texture it now simply skips all the extra logic in post and the Quest does its thing.

### CODE REVIEW

File: scene/main/viewport.cpp [hunk 1]
```diff
@@ -3186,7 +3186,7 @@ void Viewport::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "transparent_bg"), "set_transparent_background", "has_transparent_background");
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "handle_input_locally"), "set_handle_input_locally", "is_handling_input_locally");
 	ADD_GROUP("Rendering", "");
-	ADD_PROPERTY(PropertyInfo(Variant::INT, "msaa", PROPERTY_HINT_ENUM, "Disabled,2x,4x,8x,16x"), "set_msaa", "get_msaa");
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "msaa", PROPERTY_HINT_ENUM, "Disabled,2x,4x,8x,16x,External 2x,External 4x"), "set_msaa", "get_msaa");
```
Comment (BastiaanOlij) [inline scene/main/viewport.cpp]: Should we just call these Quest 2x and Quest 4x?

**Comment by BastiaanOlij** [comment]
@akien-mga you'll need to make a call on the fact we're adding a new feature after the feature freeze, I think this one is worth it :)

File: drivers/gles2/rasterizer_storage_gles2.cpp [hunk 2]
```diff
@@ -5154,16 +5161,42 @@ void RasterizerStorageGLES2::render_target_set_external_texture(RID p_render_tar
...
+			// and set our external texture as the texture...
+			glFramebufferTexture2DMultisample(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, p_texture_id, 0, msaa);
 
-		// seeing we're rendering into this directly, better also use our depth buffer, just use our existing one :)
-		if (config.support_depth_texture) {
-			glFramebufferTexture2D(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_TEXTURE_2D, rt->depth, 0);
-		} else {
-			glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_RENDERBUFFER, rt->depth);
+		} else
+#endif
```
Comment (clayjohn) [inline drivers/gles2/rasterizer_storage_gles2.cpp]: Looks like a minor whitespace issue here

File: drivers/gles2/rasterizer_storage_gles2.h [hunk 3]
```diff
@@ -1182,10 +1182,13 @@ class RasterizerStorageGLES2 : public RasterizerStorage {
 		struct External {
```
Comment (m4gr3d) [inline drivers/gles2/rasterizer_storage_gles2.h]: Unrelated to this PR, but the use of the term `External` might introduce confusion with `GL_TEXTURE_EXTERNAL_OES` textures.
Might be worth removing the confusing by renaming the struct in another PR.

File: drivers/gles2/rasterizer_storage_gles2.cpp [hunk 4]
```diff
@@ -5154,16 +5161,42 @@ void RasterizerStorageGLES2::render_target_set_external_texture(RID p_render_tar
...
-		// is there a point to setting the internal formats? we don't know them..
+		// Switch our texture on our frame buffer
+#if ANDROID_ENABLED
+		if (rt->msaa >= VS::VIEWPORT_MSAA_EXT_2X && rt->msaa <= VS::VIEWPORT_MSAA_EXT_4X) {
+			// This code only applies to the Oculus Go and Oculus Quest. Due to the the tiled nature
+			// of the GPU we can do a single render pass by rendering directly into our texture chains
+			// texture and apply MSAA as we render.
+
+			// On any other hardware these two modes are ignored and we do not have any MSAA,
+			// the normal MSAA modes need to be used to enable our two pass approach
```
Comment (m4gr3d) [inline drivers/gles2/rasterizer_storage_gles2.cpp]: @BastiaanOlij There's no actual check that prevent this from actually being used on other Android devices, correct? By this I mean if an hardware manufacturer decides to follow Oculus's approach, there's nothing preventing a plugin for that hardware to leverage this logic.
So maybe we should leave the naming and description generic enough for Android devices.

File: scene/main/viewport.cpp (hunk 1 above)
Comment (m4gr3d) [reply scene/main/viewport.cpp]: Added another comment above about naming.

**Review by m4gr3d — APPROVED** [review]

File: drivers/gles2/rasterizer_storage_gles2.cpp (hunk 4 above)
Comment (BastiaanOlij) [reply drivers/gles2/rasterizer_storage_gles2.cpp]: Nope, but it does only work on Android atm. I tried using it on desktop but it turns out there is a specific extention on Quest that allows us to do this.
I'm not sure what would be the best name, we shouldn't make it too broad or else people are going to try and use it on normal Android devices while this only works with the VR plugins.

File: drivers/gles2/rasterizer_storage_gles2.h (hunk 3 above)
Comment (BastiaanOlij) [reply drivers/gles2/rasterizer_storage_gles2.h]: Well I got there first ! :P

J/k but it is interesting, I never really got why that extension is called TEXTURE_EXTERNAL, from what I understand it simply allows the texture to be shared amongst processes so one process could be writing say camera data into it (ARCore) while another process uses it to render. What the underlying magic is that makes it special I'm not sure.

But the external texture feature in ARVR is a different thing and was added a while back, here we refer to a render texture supplied externally by a plugin. I'm all for a better name but we should apply it across the board, not just this struct but also the method call in the ARVR plugin should be renamed (seeing we use a struct of function pointers changing the name of a pointer won't have that much impact). I suggest though that we do this in Godot 4, I've been wanting to rename all the ARVR objects to XR, might aswell do a bunch of improvements in the naming of things.

File: drivers/gles2/rasterizer_storage_gles2.cpp (hunk 2 above)
Comment (BastiaanOlij) [reply drivers/gles2/rasterizer_storage_gles2.cpp]: Haven't installed clang-format yet on my new computer, time to rectify that I think :) I'll fix it up :)

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**